internal/util/chanutil/debounced_chan.go

```diff
@@ -11,45 +11,66 @@ import (
 // subsequent calls are delayed until the cooldown period has elapsed and are
 // also coalesced into a single call.
 type DebouncedChan struct {
-	done     <-chan struct{}
-	c        chan struct{}
-	cooldown time.Duration
-
-	mu                   sync.Mutex
-	timer                *time.Timer
-	sendWhenTimerExpired bool
+	c                  chan struct{}
+	cooldown           time.Duration
+	ctxDone            <-chan struct{}
+	mu                 sync.Mutex
+	sendOnTimerExpired bool
+	timer              *time.Timer
+	timerDone          bool
 }
 
 // NewDebouncedChan returns a new DebouncedChan which sends on the channel no
 // more often than the cooldown period.
 func NewDebouncedChan(ctx context.Context, cooldown time.Duration) *DebouncedChan {
 	return &DebouncedChan{
-		done:     ctx.Done(),
+		ctxDone:  ctx.Done(),
 		c:        make(chan struct{}, 1),
 		cooldown: cooldown,
 	}
 }
 
+// C is the debounced channel. Multiple invocations to Call during the cooldown
+// period will deduplicate to a single emission on this channel on the period's
+// leading edge, and one more on the trailing edge.
```

Comment:
Yeah, definitely doesn't sound like what I intended here. I wanted it to be leading edge, but if no other calls come in after the initial one nothing should happen, it should just be ready to fire again immediately on the next one.

However if a 2nd call comes in immediately after the 1st one, my thought was that we should still fire immediately a 2nd time at the end of the cooldown. And then repeat, i.e. only fire a 3rd time if yet another call comes in before `cooldown` completes (or whenever it comes in at some later time after `cooldown`).

Hopefully that makes sense? Could easily diagram it if that helps.

Reply:
kk, yeah, your explanation here is exactly how I would've intuitively expected it to work too.